### PR TITLE
Rich profile loading state

### PR DESCRIPTION
**Pull request by bfollington**
Fixes https://github.com/subconsciousnetwork/subconscious/issues/654
Fixes https://github.com/subconsciousnetwork/subconscious/issues/678
Fixes https://github.com/subconsciousnetwork/subconscious/issues/604

https://github.com/subconsciousnetwork/subconscious/assets/5009316/353b8fe8-e515-4d88-9bdc-2bd569ee9f64

- Pass partial profile data in description
- Display as much data as possible in loading state
- Add skeleton view
- Animate omnibar

# Tasks

- [x] Fade in elements rather than snapping
- [x] Memo viewer loading state
    - [x] Consider slight greywash background or text for viewer
- [x] Formatting and fixups

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Common/EmptyStateView.swift [hunk 1]
```diff
@@ -18,7 +18,7 @@ struct EmptyStateView: View {
             Spacer()
         }
         .padding(AppTheme.padding)
-        .font(.caption)
+        .font(.body)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/EmptyStateView.swift:21]: https://github.com/subconsciousnetwork/subconscious/issues/678

File: xcode/Subconscious/Shared/Components/Common/Layout/TabbedThreeColumnView.swift [hunk 2]
```diff
@@ -19,6 +19,7 @@ struct TabbedThreeColumnView<A: View, B: View, C: View>: View {
     
     var selectedColumnIndex: Int = 0
     var changeColumn: (Int) -> Void
+    var showBar: Bool = true
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Layout/TabbedThreeColumnView.swift:22]: TODO: remove this workaround,

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift [hunk 3]
```diff
@@ -11,16 +11,40 @@ import Combine
 
 struct ProfileStatisticView: View {
     var label: String
-    var count: Int
+    var count: Int?
     
     var body: some View {
         HStack(spacing: AppTheme.unit) {
-            Text("\(count)").bold()
+            Text("\(count.map { c in String(c) } ?? "-")").bold()
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift:18]: Extract to func?

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift [hunk 4]
```diff
@@ -59,29 +59,30 @@ struct UserProfileHeaderView: View {
...
-            }
+                    .font(.caption)
+                    .foregroundColor(.primary)
+                }
+            )
+            .transition(.opacity)
             
             if user.bio.hasVisibleContent {
                 Text(verbatim: user.bio.text)
+                    .transition(.opacity)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:85]: Superfluous?

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift [hunk 5]
```diff
@@ -59,29 +59,30 @@ struct UserProfileHeaderView: View {
...
+                        // ProfileStatisticView(label: "Backlinks", count: statistics.backlinkCount)
+                        ProfileStatisticView(label: "Following", count: statistics?.followingCount)
                     }
-                )
-            }
+                    .font(.caption)
+                    .foregroundColor(.primary)
+                }
+            )
+            .transition(.opacity)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:81]: Superfluous?

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift [hunk 6]
```diff
@@ -80,11 +81,12 @@ enum UserProfileDetailNotification: Hashable {
 /// profile's internal state.
 struct UserProfileDetailDescription: Hashable {
     var address: Slashlink
+    var user: UserProfile?
     var initialTabIndex: Int = UserProfileDetailModel.recentEntriesTabIndex
 }
 
 enum UserProfileDetailAction {
-    case appear(Slashlink, Int)
+    case appear(Slashlink, Int, UserProfile?)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:89]: Pass basic user information in action if we have it, allows for loading state to render a partial view.

File: xcode/Subconscious/Shared/Components/OmniboxView.swift [hunk 7]
```diff
@@ -44,15 +46,30 @@ struct OmniboxView: View {
...
+                case .loading:
+                    RoundedRectangle(cornerRadius: AppTheme.cornerRadius)
+                        .stroke(style: StrokeStyle(lineWidth: 1.5, lineCap: .round, dash: [8, 4], dashPhase: phase))
+                        .onAppear {
+                            withAnimation(.linear.repeatForever(autoreverses: false)) {
+                                phase -= 20
+                            }
+                        }
+                        .foregroundColor(.accentColor)
+                        .opacity(0.5)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/OmniboxView.swift:67]: Formatting

File: xcode/Subconscious/Shared/Components/OmniboxView.swift [hunk 8]
```diff
@@ -111,7 +128,7 @@ struct OmniboxView_Previews: PreviewProvider {
         var defaultAudience = Audience.local
 
         var body: some View {
-            OmniboxView(address: address, defaultAudience: defaultAudience)
+            OmniboxView(address: address, defaultAudience: defaultAudience, status: [.loaded, .loading].randomElement()!)
                 .onTapGesture {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/OmniboxView.swift:132]: Formatting

**Comment by justinabrahms** [comment]
That omnibar animation is :fire:

File: xcode/Subconscious/Shared/Components/OmniboxView.swift [hunk 9]
```diff
@@ -9,9 +9,11 @@ import SwiftUI
 
 struct OmniboxView: View {
     @Environment(\.colorScheme) var colorScheme
+    @State private var phase = Double.pi / 8
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/OmniboxView.swift:12]: What is the value of making this a mutable state vs an immutable property?

If mutability needed, would be nice to have a little comment detailing impl.

Non-blocking.

**Review by gordonbrander — APPROVED** [review]
Looks GREAT to me!

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift [hunk 10]
```diff
@@ -62,23 +62,21 @@ struct UserProfileHeaderView: View {
...
-                        .font(.caption)
-                        .foregroundColor(.primary)
+            Button(
+                action: {
+                    onTapStatistics()
+                },
+                label: {
+                    HStack(spacing: AppTheme.unit2) {
+                        ProfileStatisticView(label: "Notes", count: statistics?.noteCount)
+                        // TODO: put this back when we have backlink count
```
[the remark this answers is not in the document]
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:72]: 👍 https://github.com/subconsciousnetwork/subconscious/issues/689

**Pull request merged**